src/vs/editor/contrib/linesOperations/linesOperations.ts

```diff
@@ -112,18 +112,74 @@ abstract class AbstractMoveLinesAction extends EditorAction {
 		let commands: ICommand[] = [];
 		let selections = editor.getSelections() || [];
 		let autoIndent = editor.getConfiguration().autoIndent;
+		let movingMultipleLines = false;
+		let selectionDirectionDown = true;
+		let newSelections: Selection[] = [];
+
+		if (selections.length > 1) {
+			movingMultipleLines = selections[0].endLineNumber !== selections[selections.length - 1].endLineNumber;
+
+			if (movingMultipleLines) {
+				// Stash selections while processing, set new selections +/- line-change only
+				for (const selection of selections) {
+					let startCol = selection.getStartPosition().column;
+					let startLine = selection.getStartPosition().lineNumber;
+					let newStartLine = this.down ? startLine + 1 : startLine - 1;
+
+					let endCol = selection.getEndPosition().column;
+					let endLine = selection.getEndPosition().lineNumber;
+					let newEndLine = this.down ? endLine + 1 : endLine - 1;
+
+					newSelections.push(new Selection(newStartLine, startCol, newEndLine, endCol));
+				}
+
+				// Work only with one selection per line
```

Comment:
Let's put this filter operation before the for loop above, so as not to loop over uneccessary selections in the for loop. 

Also instead of doing a filter operation followed by a map operation followed by an indexOf operation which is O(n^2) where `n` is the number of selections could we just loop once over the selections, keep track of the `endLineNumber` of the last processed selection and if the current selection does not match this `endLineNumber` add the current selection to a pre-created selections array. This will allow us to reduce the complexity to O(n).